research/fed-sm/job_configs/fedsm_prostate/app/custom/learners/supervised_monai_prostate_fedsm_learner.py

```diff
@@ -0,0 +1,392 @@
+            aggregation_epochs: the number of training epochs of both global and personalized models for a round. Defaults to 1.
+            train_task_name: name of the task to train the model.
+
+        Returns:
+            a Shareable with the updated local model after running `execute()`
+        """
+        SupervisedMonaiProstateLearner.__init__(
+            self,
+            train_config_filename=train_config_filename,
+            aggregation_epochs=aggregation_epochs,
+            train_task_name=train_task_name,
+        )
+        self.fedsm_person_model_epochs = aggregation_epochs
+        self.fedsm_select_model_epochs = fedsm_select_epochs
+
+    def train_config(self, fl_ctx: FLContext):
+        # Initialize superclass
+        SupervisedMonaiProstateLearner.train_config(self, fl_ctx)
+
+        engine = fl_ctx.get_engine()
+        ws = engine.get_workspace()
+        app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+
+        # Initialize PTFedSMHelper
+        # personalized and selector model training epoch
+        # personalized model same as global model
+        # selector model can be different from the other two task models
+        fedsm_person_model = UNet(
+            spatial_dims=2,
+            in_channels=1,
+            out_channels=1,
+            channels=(16, 32, 64, 128, 256),
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        ).to(self.device)
+        fedsm_select_model = vgg11(
+            num_classes=self.config_info["select_num_classes"],
+        ).to(self.device)
+        fedsm_person_criterion = DiceLoss(sigmoid=True)
+        fedsm_select_criterion = torch.nn.CrossEntropyLoss()
+        fedsm_person_optimizer = optim.Adam(fedsm_person_model.parameters(), lr=self.lr)
+        fedsm_select_optimizer = optim.Adam(
+            fedsm_select_model.parameters(), lr=self.config_info["learning_rate_select"]
+        )
+        self.fedsm_helper = SupervisedPTFedSMHelper(
+            person_model=fedsm_person_model,
+            select_model=fedsm_select_model,
+            person_criterion=fedsm_person_criterion,
+            select_criterion=fedsm_select_criterion,
+            person_optimizer=fedsm_person_optimizer,
+            select_optimizer=fedsm_select_optimizer,
+            device=self.device,
+            app_dir=app_dir,
+            person_model_epochs=self.fedsm_person_model_epochs,
+            select_model_epochs=self.fedsm_select_model_epochs,
+        )
+
+    def train(
+        self,
+        shareable: Shareable,
+        fl_ctx: FLContext,
+        abort_signal: Signal,
+    ) -> Shareable:
+        """Training task pipeline for FedSM
+        Get global/client/selector model weights (potentially with HE)
+        Prepare for fedprox loss - general
+        Local training all three models
+        Return updated weights of all three models (model_diff)
+        """
+        if abort_signal.triggered:
```

Comment:
I would remove the if abort_signal.triggered checks in train(). Should be sufficient if the sub functions check it, e.g. `local_train_select`

Reply:
You need to `return make_reply(ReturnCode.TASK_ABORTED)` if it has been triggered. Check once in the end.

Reply:
removed